File: vi/codes/swift/chapter_dynamic_programming/climbing_stairs_dfs.swift
/**
 * File: climbing_stairs_dfs.swift
 * Created Time: 2023-07-15
 * Author: nuomi1 (nuomi1@qq.com)
 */

/* Tìm kiếm */
func dfs(i: Int) -> Int {
    // Đã biết dp[1] và dp[2], trả về chúng
    if i == 1 || i == 2 {
        return i
    }
    // dp[i] = dp[i-1] + dp[i-2]
    let count = dfs(i: i - 1) + dfs(i: i - 2)
    return count
}

/* Leo cầu thang: tìm kiếm */
func climbingStairsDFS(n: Int) -> Int {
    dfs(i: n)
}

@main
enum ClimbingStairsDFS {
    /* Mã điều khiển */
    static func main() {
        let n = 9

        let res = climbingStairsDFS(n: n)
        print("Climbing \(n) stairs has \(res) solutions")
    }
}

File: vi/codes/typescript/chapter_dynamic_programming/climbing_stairs_dfs.ts
/**
 * File: climbing_stairs_dfs.ts
 * Created Time: 2023-07-26
 * Author: yuan0221 (yl1452491917@gmail.com)
 */

/* Tìm kiếm */
function dfs(i: number): number {
    // Đã biết dp[1] và dp[2], trả về chúng
    if (i === 1 || i === 2) return i;
    // dp[i] = dp[i-1] + dp[i-2]
    const count = dfs(i - 1) + dfs(i - 2);
    return count;
}

/* Leo cầu thang: Tìm kiếm */
function climbingStairsDFS(n: number): number {
    return dfs(n);
}

/* Driver Code */
const n = 9;
const res = climbingStairsDFS(n);
console.log(`Climbing ${n} stairs has ${res} solutions`);

export {};
